src/lib/io/sol-memmap-storage.c

```diff
@@ -251,15 +301,151 @@ check_version(const struct sol_memmap_map *map)
     return sol_ptr_vector_append(&checked_maps, (void *)map) == 0;
 }
 
+static struct map_timeout *
+get_map_timeout(const struct sol_memmap_map *map)
+{
+    struct map_timeout *map_timeout;
+    int i;
+
+    SOL_VECTOR_FOREACH_IDX (&write_timeouts, map_timeout, i) {
+        if (map_timeout->map == map)
+            return map_timeout;
+    }
+
+    assert(false); /* Should never get here */
+
+    return NULL;
+}
+
+static bool
+perform_pending_writes(void *data)
+{
+    int i, r;
+    struct pending_write_data *pending;
+    FILE *file = NULL;
+    const struct sol_memmap_map *map = data;
+    struct map_timeout *map_timeout = get_map_timeout(map);
+
+    map_timeout->timeout = NULL;
+
+    file = fopen(map->path, "r+e");
+    if (!file) {
+        SOL_WRN("Error opening file [%s]: %s", map->path,
+            sol_util_strerrora(errno));
+        return false;
+    }
+
+    SOL_VECTOR_FOREACH_IDX (&map_timeout->pending_writes, pending, i) {
+        sol_memmap_write_raw_do(map->path, pending->name, pending->entry,
```

Comment:
I think this could call a `pending->cb` that may modify `map_timeout->pending_writes`, causing a bug here. Since we're going to clear the vector later, I suggest you steal that vector and reset the old pointer:

```c
struct sol_vector tmp_vector = map_timeout->pending_writes;
sol_vector_init(&map_timeout->pending_writes, sizeof(struct pending_write_data));

SOL_VECTOR_FOREACH_IDX (&tmp_vector, ...) {...}
sol_vector_clear(&tmp_vector);
```